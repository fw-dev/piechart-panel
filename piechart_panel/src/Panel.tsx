import React from 'react';
// @ts-ignore
import Chart from 'chart.js';
// @ts-ignore
import TimeSeries from './TimeSeries';

import { State, Props } from './types';
import { defaultChartConfig, initialState, colors, defaultHighlight } from './defaults';
import 'css/filewave-piechart-panel.css';

export class Panel extends React.Component<Props, State> {
  panelId: number | undefined = this.props.data.request && this.props.data.request.panelId;
  chart: any;
  state = initialState;

  componentDidMount() {
    this.handleDataFormatting(this.props.data);
  }

  componentDidUpdate(prevProps: Props, prevState: State) {
    const { options, data } = this.props;
    const { highlight, chartData } = this.state;

    if (options !== prevProps.options) {
      this.updateChart();
    }

    if (highlight.label !== prevState.highlight.label || options.highlightEnabled !== prevProps.options.highlightEnabled) {
      this.drawChart();
    }

    if (data !== prevProps.data || options.aliasColors !== prevProps.options.aliasColors) {
      this.handleDataFormatting(data);
    }

    if (chartData !== prevState.chartData) {
      this.updateHighlight();
    }
  }

  updateHighlight = () => {
    const { selectedHighlight } = this.props.options;
    const { highlightData } = this.state;
    const highlight = highlightData.find(highlight => highlight.label === selectedHighlight) || defaultHighlight;
    this.setState({ highlight });
  };

  handleDataFormatting = (data: any) => {
    const timeSeries = data.series.map((serie: any) => this.createTimeSeries(serie));
    const chartData = this.formatChartData(timeSeries, data.series);
    const highlightData = this.formatHighlightData(timeSeries);
    this.setState({
      chartData,
      highlightData,
    });
  };

  formatHighlightData = (timeSeries: any) => {
    const { options } = this.props;
    const total = timeSeries.reduce((x: number, y: any) => x + y.stats.total, 0);
    const { highlightValue } = this.props.options;
    const highlightData = timeSeries.map((serie: any) => {
      const percentage = `${(serie.stats[options.valueName] / (total / 100) || 0).toFixed()}%`;
      const value = serie.stats[options.valueName];
      return {
        label: serie.label,
        value: highlightValue === 'percentage' ? percentage : value,
      };
    });

    return highlightData;
  };

  formatChartData = (timeSeries: any, series: any) => {
    const { valueName, aliasColors } = this.props.options;
    const chartData = {
      labels: timeSeries.map((serie: any) => serie.alias),
      datasets: [
        {
          data: timeSeries.map((serie: any) => serie.stats[valueName]),
          backgroundColor: timeSeries.map((serie: any, i: number) => aliasColors[serie.alias] || colors[i]),
          metadata: series.map((serie: any) => serie.fields.find((field: any) => field.labels).labels),
        },
      ],
    };

    return chartData;
  };

  createTimeSeries = (serie: any) => {
    const timeSeries = new TimeSeries({
      datapoints: [serie.fields.map((field: any) => field.values.buffer[0])],
      alias: serie.name,
      target: serie.name,
    });
    timeSeries.flotpairs = timeSeries.getFlotPairs(this.props.options.nullPointMode);

    return timeSeries;
  };

  updateChartSettings = () => {
    const { options, data } = this.props;
    return {
      ...defaultChartConfig,
      data: this.state.chartData,
      type: options.chartType,
      plugins: [
        {
          afterDraw: () => {
            if (!data.series.length) {
              return this.drawDataUnavailableMessage();
            }
            if (options.highlightEnabled && options.chartType === 'doughnut') {
              return this.drawHighlight();
            }
          },
        },
      ],
      options: {
        ...defaultChartConfig.options,
        cutoutPercentage: options.chartType === 'doughnut' ? parseInt(options.cutoutPercentage, 0) : 0,
        onClick: this.handleClick,
        legend: {
          display: options.legendEnabled,
          position: options.legendPosition,
          align: options.legendAlign,
          labels: {
            boxWidth: parseInt(options.legendBoxWidth, 0),
            fontSize: parseInt(options.legendFontSize, 0),
            usePointStyle: options.legendUsePointStyle,
          },
        },
      },
    };
  };

  generateUrl = (target: any) => {
    const { linkUrl } = this.props.options;
    let url = linkUrl;
    const variableRegex = /(\${__)(.*?)(})/g; // Match anything that's written between ${__...}
    const queryVariables = Object.keys(target.metadata).map((variable: string) => ({ name: variable, value: target.metadata[variable] }));
    const vars = [...queryVariables];
    // @ts-ignore, need to look into fixing "Object is possibly null" here
    const matchedVars = linkUrl.match(variableRegex).map((variable: string) => {
      const matchedVariable = vars.find((item: any) => item.name === variable.substring(4, variable.length - 1)) || { value: '', name: '' };
      return {
        name: variable,
        value: variable.includes('targetLabel') ? target.label : variable.includes('targetValue') ? target.value : matchedVariable.value || variable,
      };
    });
    matchedVars.forEach((match: any) => (url = url.replace(match.name, match.value)));

    return url;
  };

  handleClick = (_event: Event, targets: any) => {
    const { options } = this.props;
    if (!options.linkEnabled || !targets.length) {
      return;
    }

    const { labels, datasets } = this.chart.data;
    const targetIndex = targets[0]._index;
    const url = this.generateUrl({
      label: labels[targetIndex],
      value: datasets[0].data[targetIndex],
      metadata: datasets[0].metadata[targetIndex],
    });

    return window.open(url, options.linkTargetBlank ? '_blank' : 'currentWindow');
  };

  updateChart = () => {
    const { options } = this.props;
    if (this.chart.config.type !== options.chartType) {
      return this.drawChart();
    }
    if (options.highlightEnabled && options.selectedHighlight) {
      this.updateHighlight();
    }
    this.chart.options = this.updateChartSettings().options;
    this.chart.update({ duration: 0 });
  };

  drawDataUnavailableMessage = () => {
    const {
      ctx,
      chartArea: { right, bottom },
    } = this.chart.chart;
    const xPos = right;
    const yPos = bottom / 2;
    this.drawText(ctx, this.props.options.dataUnavailableMessage, xPos, yPos, 18);
  };

  drawHighlight = () => {
    const {
      ctx,
      width,
      height,
      chartArea: { left, right, top, bottom },
    } = this.chart.chart;
    const { position } = this.chart.legend;
    const { label, value } = this.state.highlight;
    const xPos = Math.round(position === 'left' ? width + left : right);
    const yPos = (position === 'top' ? height + top : bottom) / 2;
    this.drawText(ctx, value, xPos, yPos - 5, 32);
    this.drawText(ctx, label, xPos, yPos + 25, 18);
  };

  drawText = (ctx: any, value: any, xPos: number, yPos: number, fontSize: number) => {
    ctx.restore();
    ctx.textBaseline = 'middle';
    ctx.font = `${fontSize}px sans-serif`;
    ctx.fillStyle = '#BABABA';
    xPos = (xPos - ctx.measureText(value).width) / 2;
    ctx.fillText(value, xPos, yPos);
    ctx.save();
  };

  drawChart = () => {
    const canvas: HTMLElement | HTMLCanvasElement | null = document.getElementById(`fw-piechart-${this.panelId}`);
    if (this.chart && this.chart.id >= 0) {
      // If a chart exists, destroy it, otherwise it will add another one to the canvas
      this.chart.destroy();
    }

    if (canvas instanceof HTMLCanvasElement) {
      this.chart = new Chart(canvas.getContext('2d'), { ...this.updateChartSettings() });
    }
  };

  render() {
    const { width, height } = this.props;
    return (
      <div className="fw-piechart" style={{ width, height }}>
        <canvas id={`fw-piechart-${this.panelId}`} />
      </div>
    );
  }
}
