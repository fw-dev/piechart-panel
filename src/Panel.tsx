import React, { useRef, useEffect, useState } from 'react';
import { PanelProps } from '@grafana/data';
// @ts-ignore
import Chart from 'chart.js';

import TimeSeries from 'grafana/app/core/time_series2';
import { PanelOptions } from 'types';
import { defaultChartConfig, defaultChartData, defaultHighlight, colors } from 'defaults';
import 'css/filewave-piechart-panel.css';

interface Props extends PanelProps<PanelOptions> {}
let chart: any = null;

export const Panel = ({ options, data, width, height }: Props) => {
  const chartRef = useRef(null);
  const [chartData, setChartData] = useState(defaultChartData);
  const [highlight, setHighlight] = useState(defaultHighlight);

  useEffect(() => {
    // Mounting
    drawChart();
  }, []);

  useEffect(() => {
    redrawChart();
  }, [options]);

  useEffect(() => {
    drawChart();
  }, [highlight, options.highlightEnabled]);

  useEffect(() => {
    handleDataChange(data);
  }, [data, options.aliasColors]);

  useEffect(() => {
    drawChart();
    updateHighlight();
  }, [chartData]);

  const updateHighlight = () => {
    const { selectedHighlight, highlightValue } = options;
    const labels = chartData.labels;
    const chartValues = chartData.datasets[0].data || [];
    const index = labels.indexOf(selectedHighlight);
    const total = chartValues.reduce((x, y) => x + y, 0);
    const value = chartValues[index] || 0;
    const percentage = `${(value / (total / 100) || 0).toFixed()}%`;

    setHighlight({
      value: highlightValue === 'percentage' ? percentage : value.toString(),
      label: selectedHighlight,
    });
  };

  const handleDataChange = (data: any) => {
    const timeSeries = data.series.map((serie: any) => createTimeSeries(serie));
    const updatedData = formatChartData(timeSeries, data.series);
    setChartData(updatedData);
    drawChart();
  };

  const formatChartData = (timeSeries: any, series: any) => {
    const { aliasColors } = options;
    const chartData = {
      labels: timeSeries.map((serie: any) => serie.alias),
      datasets: [
        {
          data: timeSeries.map((serie: any) => serie.stats[options.valueName]),
          backgroundColor: timeSeries.map((serie: any, i: number) => aliasColors[serie.alias] || colors[i]),
          metadata: series.map((serie: any) => serie.fields.find((field: any) => field.labels).labels),
        },
      ],
    };

    return chartData;
  };

  const createTimeSeries = (serie: any) => {
    const timeSeries = new TimeSeries({
      datapoints: [serie.fields.map((field: any) => field.values.buffer[0])],
      alias: serie.name,
      target: serie.name,
    });
    timeSeries.flotpairs = timeSeries.getFlotPairs(options.nullPointMode);

    return timeSeries;
  };

  const updateChartSettings = () => ({
    ...defaultChartConfig,
    data: chartData,
    type: options.chartType,
    plugins: [
      {
        afterDraw: () => {
          if (!data.series.length) {
            return drawDataUnavailableMessage();
          }
          if (options.highlightEnabled && options.chartType === 'doughnut') {
            return drawHighlight();
          }
        },
      },
    ],
    options: {
      ...defaultChartConfig.options,
      cutoutPercentage: options.chartType === 'doughnut' ? parseInt(options.cutoutPercentage, 0) : 0,
      onClick: handleClick,
      legend: {
        display: options.legendEnabled,
        position: options.legendPosition,
        align: options.legendAlign,
        labels: {
          boxWidth: parseInt(options.legendBoxWidth, 0),
          fontSize: parseInt(options.legendFontSize, 0),
          usePointStyle: options.legendUsePointStyle,
        },
      },
    },
  });

  const generateUrl = (target: any) => {
    const { linkUrl } = options;
    let url = linkUrl;
    const variableRegex = /(\${__)(.*?)(})/g; // Match anything that's written between ${__...}
    const queryVariables = Object.keys(target.metadata).map((variable: string) => ({ name: variable, value: target.metadata[variable] }));
    const vars = [...queryVariables];
    // @ts-ignore, need to look into fixing "Object is possibly null" here
    const matchedVars = linkUrl.match(variableRegex).map((variable: string) => {
      const matchedVariable = vars.find((item: any) => item.name === variable.substring(4, variable.length - 1)) || { value: '', name: '' };
      return {
        name: variable,
        value: variable.includes('targetLabel') ? target.label : variable.includes('targetValue') ? target.value : matchedVariable.value || variable,
      };
    });
    matchedVars.forEach((match: any) => (url = url.replace(match.name, match.value)));

    return url;
  };

  const handleClick = (_event: any, targets: any) => {
    if (!options.linkEnabled || !targets.length) {
      return;
    }

    const { labels, datasets } = chart.data;
    const targetIndex = targets[0]._index;
    const url = generateUrl({
      label: labels[targetIndex],
      value: datasets[0].data[targetIndex],
      metadata: datasets[0].metadata[targetIndex],
    });

    return window.open(url, options.linkTargetBlank ? '_blank' : 'currentWindow');
  };

  const redrawChart = () => {
    if (chart.config.type !== options.chartType) {
      return drawChart();
    }
    if (options.highlightEnabled && options.selectedHighlight) {
      updateHighlight();
    }
    chart.options = updateChartSettings().options;
    chart.update({ duration: 0 });
  };

  const drawDataUnavailableMessage = () => {
    const { ctx } = chart.chart;
    const { right, bottom } = chart.chart.chartArea;
    const xPos = right;
    const yPos = bottom / 2;
    drawText(ctx, options.dataUnavailableMessage, xPos, yPos, 18);
  };

  const drawHighlight = () => {
    const { ctx } = chart.chart;
    const { position } = chart.legend;
    const { left, right, top, bottom } = chart.chart.chartArea;
    const { width, height } = chart.chart;
    const { label, value } = highlight;
    const xPos = Math.round(position === 'left' ? width + left : right);
    const yPos = (position === 'top' ? height + top : bottom) / 2;
    drawText(ctx, value, xPos, yPos - 5, 32);
    drawText(ctx, label, xPos, yPos + 25, 18);
  };

  const drawText = (ctx: any, value: string, xPos: number, yPos: number, fontSize: number) => {
    ctx.restore();
    ctx.textBaseline = 'middle';
    ctx.font = `${fontSize}px sans-serif`;
    ctx.fillStyle = '#BABABA';
    xPos = (xPos - ctx.measureText(value).width) / 2;
    ctx.fillText(value, xPos, yPos);
    ctx.save();
  };

  const drawChart = () => {
    if (chart !== null) {
      // Destroy the chart, otherwise it will add another one to the canvas
      chart.destroy();
    }
    chart = new Chart(chartRef.current, { ...updateChartSettings() });
  };

  return (
    <div className="fw-piechart" style={{ width, height }}>
      <canvas ref={chartRef} />
    </div>
  );
};
