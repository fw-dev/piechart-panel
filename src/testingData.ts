import { defaultPanelOptions } from './defaults';

export const testData = (state?: string) => ({
  state: state || 'Done',
  error: undefined,
  timeRange: {},
  request: {
    cacheTimeout: null,
    dashboardId: 11,
    endTime: 1577184404002,
    interval: '1m',
    intervalMs: 60000,
    maxDataPoints: 449,
    panelId: 2,
    rangeRaw: {
      from: 'now-6h',
      to: 'now',
    },
    requestId: 'Q100',
    startTime: 1577184403903,
    scopedVars: {
      __interval: {
        text: '1m',
        value: '1m',
      },
      __interval_ms: {
        text: '60000',
        value: 60000,
      },
    },
    range: {},
    targets: [
      {
        datasource: 'FileWave Prometheus',
        expr: 'go_memstats_alloc_bytes_total',
        instant: true,
        legendFormat: '{{job}}',
        refId: 'A',
        requestId: '2A',
      },
    ],
    timeInfo: '',
    timezone: 'browser',
  },
  series: [
    {
      fields: [
        {
          calcs: null,
          config: { unit: undefined },
          labels: { __name__: 'go_memstats_alloc_bytes_total', instance: 'localhost:20448', job: 'grafana' },
          name: 'grafana',
          type: 'number',
          values: {
            buffer: [1078024840],
          },
        },
      ],
      length: 1,
      meta: undefined,
      name: 'grafana',
      refId: 'A',
    },
  ],
});

export const defaultProps = {
  id: 1,
  data: {
    ...testData(),
  },
  timeRange: '',
  timeZone: '',
  options: {
    ...defaultPanelOptions,
    aliasColors: {},
  },
  onOptionsChange: jest.fn(),
  renderCounter: null,
  transparent: false,
  width: 500,
  height: 500,
  replaceVariables: jest.fn(),
  onChangeTime: jest.fn(),
  onChangeTimeRange: jest.fn(),
};
