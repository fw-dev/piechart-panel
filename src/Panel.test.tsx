import React from 'react';
import 'jest-canvas-mock';
// @ts-ignore
import { configure, mount } from 'enzyme';
// @ts-ignore
import Adapter from 'enzyme-adapter-react-16';
import { Panel } from './Panel';
import { defaultProps, testData } from './testingData';

configure({ adapter: new Adapter() });

jest.mock('grafanaUtils', () => {
  return {
    TimeSeries: jest.fn().mockImplementation(() => ({
      getFlotPairs: jest.fn(),
      stats: {},
    })),
    kbn: jest.fn(),
  };
});

describe('Piechart Panel', () => {
  describe('when doing a basic render with data', () => {
    // Arrange
    const component = <Panel {...defaultProps} />;

    // Act
    const mounted = mount(component);

    // Assert
    it('should render correctly', () => {
      expect(mounted).toMatchSnapshot();
    });
  });

  describe('when the panel options are being updated', () => {
    // Arrange
    const component = <Panel {...defaultProps} />;
    const mounted = mount(component);
    const instance = mounted.instance();
    const updateChart = jest.fn();
    instance.updateChart = updateChart;

    // Act
    mounted.setProps({
      ...defaultProps,
      options: {
        ...defaultProps.options,
        chartType: 'doughnut',
      },
    });

    // Assert
    it('should call the updateChart method', () => {
      expect(updateChart).toHaveBeenCalled();
    });
  });

  describe('when the incoming data is being updated', () => {
    // Arrange
    const component = <Panel {...defaultProps} />;
    const mounted = mount(component);
    const instance = mounted.instance();
    const handleDataFormatting = jest.fn();
    instance.handleDataFormatting = handleDataFormatting;

    // Act
    mounted.setProps({
      ...defaultProps,
      data: testData('Loading'),
    });

    // Assert
    it('should call the handleDataFormatting method', () => {
      expect(handleDataFormatting).toHaveBeenCalled();
    });
  });
});
