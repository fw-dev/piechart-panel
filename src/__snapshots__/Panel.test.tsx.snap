// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Piechart Panel when doing a basic render with data should render correctly 1`] = `
<Panel
  data={
    Object {
      "error": undefined,
      "request": Object {
        "cacheTimeout": null,
        "dashboardId": 11,
        "endTime": 1577184404002,
        "interval": "1m",
        "intervalMs": 60000,
        "maxDataPoints": 449,
        "panelId": 2,
        "range": Object {},
        "rangeRaw": Object {
          "from": "now-6h",
          "to": "now",
        },
        "requestId": "Q100",
        "scopedVars": Object {
          "__interval": Object {
            "text": "1m",
            "value": "1m",
          },
          "__interval_ms": Object {
            "text": "60000",
            "value": 60000,
          },
        },
        "startTime": 1577184403903,
        "targets": Array [
          Object {
            "datasource": "FileWave Prometheus",
            "expr": "go_memstats_alloc_bytes_total",
            "instant": true,
            "legendFormat": "{{job}}",
            "refId": "A",
            "requestId": "2A",
          },
        ],
        "timeInfo": "",
        "timezone": "browser",
      },
      "series": Array [
        Object {
          "fields": Array [
            Object {
              "calcs": null,
              "config": Object {
                "unit": undefined,
              },
              "labels": Object {
                "__name__": "go_memstats_alloc_bytes_total",
                "instance": "localhost:20448",
                "job": "grafana",
              },
              "name": "grafana",
              "type": "number",
              "values": Object {
                "buffer": Array [
                  1078024840,
                ],
              },
            },
          ],
          "length": 1,
          "meta": undefined,
          "name": "grafana",
          "refId": "A",
        },
      ],
      "state": "Done",
      "timeRange": Object {},
    }
  }
  height={500}
  id={1}
  onChangeTime={[MockFunction]}
  onChangeTimeRange={[MockFunction]}
  onOptionsChange={[MockFunction]}
  options={
    Object {
      "aliasColors": Object {},
      "chartType": Object {
        "label": "Pie",
        "value": "pie",
      },
      "cutoutPercentage": "80",
      "dataUnavailableMessage": "Data unavailable",
      "format": Object {
        "label": "short",
        "value": "short",
      },
      "highlightEnabled": true,
      "highlightValue": Object {
        "label": "Percentage",
        "value": "percentage",
      },
      "legendAlign": Object {
        "label": "Center",
        "value": "center",
      },
      "legendBoxWidth": "40",
      "legendEnabled": true,
      "legendFontSize": "12",
      "legendPosition": Object {
        "label": "Left",
        "value": "left",
      },
      "legendUsePointStyle": false,
      "linkEnabled": false,
      "linkTargetBlank": true,
      "linkUrl": "",
      "nullPointMode": "connected",
      "selectedHighlight": Object {
        "label": "",
        "value": "",
      },
      "valueName": "current",
    }
  }
  renderCounter={null}
  replaceVariables={[MockFunction]}
  timeRange=""
  timeZone=""
  transparent={false}
  width={500}
>
  <div
    className="fw-piechart"
    style={
      Object {
        "height": 500,
        "width": 500,
      }
    }
  >
    <canvas
      id="fw-piechart-2"
    />
  </div>
</Panel>
`;
